community/primitive-collections/src/main/java/org/neo4j/concurrent/RecentK.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.concurrent;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Tracks an (approximate) set of recently seen unique elements in a stream, based on a concurrent LRU implementation.
+ *
+ * @param <Type> the entry type stored
+ */
+public class RecentK<Type> implements Iterable<Type>
+{
+    private final int maxItems;
+
+    /**
+     * Maps items to their slots in the LRU queue, the keys in here is what we use as the source of truth
+     * for what are the most recent items.
+     */
+    private final ConcurrentHashMap<Type, Slot> index = new ConcurrentHashMap<>();
+
+    /** Most recently seen item */
+    // guarded by synchronized
+    private Slot head;
+
+    /** Least recently used item */
+    // guarded by synchronized
+    private Slot tail;
```

Comment:
The mechanism of locking is less important than the object that is the lock. I'd write "guarded by 'this'" or "guarded by synchronized(this)".